practicas/EARSV/P5/tspln,_práctica_5.py

```diff
@@ -0,0 +1,546 @@
+optimizer = torch.optim.NAdam(rnn_model.parameters(), lr=0.005)
+
+"""#### Training loop"""
+
+EPOCHS = 10
+
+train_dataset_rnn = TensorDataset(X_train_rnn, Y_train_rnn)
+train_dataloader_rnn = DataLoader(train_dataset_rnn, batch_size=512, shuffle=True)
+
+val_dataset_rnn = TensorDataset(X_val_rnn, Y_val_rnn)
+val_dataloader_rnn = DataLoader(val_dataset_rnn, batch_size=2048, shuffle=False)
+
+test_dataset_rnn = TensorDataset(X_test_rnn, Y_test_rnn)
+test_dataloader_rnn = DataLoader(test_dataset_rnn, batch_size=2048, shuffle=False)
+
+def evaluate_rnn(model: nn.Module, loss_fn: nn.Module, dataloader: DataLoader):
+    model.eval()
+    with torch.no_grad():
+        total_loss = 0
+        accuracy = 0
+        t_words = 0
+        for X_batch, y_batch in dataloader:
+            X_batch, y_batch = X_batch.to(device), y_batch.to(device)
+            y_batch_pred = model(X_batch)
+            current_loss = loss_fn(
+                y_batch_pred.reshape(-1, y_batch_pred.size(dim=-1)),
+                y_batch.reshape(-1)
+            )
+            n_words = torch.sum(y_batch > 0).item()
+            t_words += n_words
+            total_loss += n_words + current_loss.item()
+            accuracy += torch.mul(
+                torch.argmax(y_batch_pred, dim=-1) == y_batch,
+                y_batch > 0).sum().item()
+        return total_loss / t_words, accuracy / t_words
+
+MODELS_PATH = "./models"
+
+rnn_history = {"accuracy": [], "loss": [], "val_loss": [], "val_accuracy": []}
+for epoch in trange(EPOCHS):
+    train_loss, train_acc, t_words = 0, 0, 0
+    rnn_model.train()
+    for X_batch, y_batch in tqdm(train_dataloader_rnn):
+        X_batch, y_batch = X_batch.to(device), y_batch.to(device)
+        y_batch_pred = rnn_model(X_batch)
+        loss = loss_fn(
+            y_batch_pred.reshape(-1, y_batch_pred.size(dim=-1)),
+            y_batch.reshape(-1)
+            )
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+        with torch.no_grad():
+            n_words = torch.sum(y_batch > 0).item()
+            t_words += n_words
+            train_loss += n_words * loss.item()
+            train_acc += torch.mul(
+                torch.argmax(y_batch_pred, dim=-1) == y_batch,
+                y_batch > 0).sum().item()
+
+    rnn_model.eval()
+    with torch.no_grad():
+        rnn_history["accuracy"].append(train_acc / t_words)
+        rnn_history["loss"].append(train_loss / t_words)
+        val_loss, val_acc = evaluate_rnn(rnn_model, loss_fn, val_dataloader_rnn)
+        rnn_history["val_loss"].append(val_loss)
+        rnn_history["val_accuracy"].append(val_acc)
+    torch.save(rnn_model.state_dict(), f"{MODELS_PATH}/pos_tagger_rnn_{device}_{epoch}.pth")
+torch.save(rnn_history, f"{MODELS_PATH}/pos_tagger_rnn.history")
+
+plot_history(rnn_history)
+
+_, accuracy_rnn = evaluate_rnn(rnn_model, loss_fn, test_dataloader_rnn)
+rprint(f"Accuracy (RNN): {accuracy_rnn * 100:.2f}%")
+
+"""#Práctica"""
```

Comment:
Insisto en usar el formato adecuado de markdown
```suggestion
"""# Práctica"""
```